Scale parameters

teunbrand opened the pull request:
This PR follows up on #4271 and #5444, aiming to fix #4269.

Like before, it aims to provide the convenience of setting some scale parameters, without comitting to a full scale yet.
Unlike #5444, this PR is closer to Claus' approach in #4271.

This PR the `scale_params()` function that populates a new `ScalesList$params` field. When a new scale is added, it will get updated with the parameters in that field. The update mechanism required the most attention. Every `<Scale>` class now has a `Scale$new()` method that takes on most of the work that `continuous_scale()`, `discrete_scale()`, `binned_scale()` constructors were doing before. The `Scale$update()` method is a thin wrapper around this and should resolve gnarly situations (e.g. limit transformation).

In the demo below we can note:
* the y-limits are appropriately transformed by the sqrt transformation. If we were to use the transform first and then the limits, the result would be the same.
* We can set parameters for multiple scales at once, like colour/shape.

``` r
devtools::load_all("~/packages/ggplot2/")
#> ℹ Loading ggplot2
p <- ggplot(mpg, aes(displ, hwy, colour = drv, shape = factor(year))) +
  geom_point()

p +
  scale_params("x", breaks = c(2, 4, 7)) +
  scale_params("y", limits = c(0, 50)) +
  scale_params("y", transform = "sqrt") +
  scale_params(c("colour", "shape"), guide = guide_legend(ncol = 2))
```

![](https://i.imgur.com/JZRdicL.png)<!-- -->

The update mechanism should also recognise invalid updates:

``` r
p + scale_params("colour", foobar = "i dont exist")
#> Error in `geom_point()`:
#> ! Problem while computing aesthetics.
#> ℹ Error occurred in the 1st layer.
#> Caused by error in `scale_colour_discrete()`:
#> ! Cannot update scale with the unknown argument:
#> • "foobar"
```

<sup>Created on 2024-11-22 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

teunbrand committed:
use scale class constructors

teunbrand committed:
update mechanism

teunbrand committed:
user-facing function

teunbrand committed:
document

teunbrand committed:
resolve merge conflict

Merge branch 'main' into scale_experiment

# Conflicts:
#	R/scale-.R

teunbrand committed:
restore #5933

teunbrand committed:
fix vectorised aesthetics case

teunbrand committed:
protect against invalid arguments

teunbrand committed:
fix bug where limits is transformed every time

teunbrand committed:
add test

thomasp85 commented on R/scale-.R:
```diff
@@ -633,6 +547,29 @@ check_breaks_labels <- function(breaks, labels, call = NULL) {
 }
 
+check_update_params <- function(scale, params) {
+  if (inherits(scale, "ScaleContinuous")) {
+    args <- fn_fmls_names(continuous_scale)
+  } else if (inherits(scale, "ScaleDiscrete")) {
+    args <- fn_fmls_names(discrete_scale)
+  } else if (inherits(scale, "ScaleBinned")) {
+    args <- fn_fmls_names(binned_scale)
```
Would be nice if this information was somehow incorporated into the class, rather than having to check an "unlinked" constructor function

teunbrand committed:
fix merge conflict

Merge branch 'main' into scale_experiment

# Conflicts:
#	R/scale-.R
#	R/scales-.R
#	tests/testthat/test-scales.R

teunbrand committed:
use field to store params to update

teunbrand committed:
add test for re-transforming

teunbrand committed:
add test for consistency of `updatable params` field

teunbrand commented:
This could be a nice editioning carrot

teunbrand committed:
resolve merge conflict

Merge branch 'main' into scale_experiment

# Conflicts:
#	NAMESPACE
#	R/plot-construction.R
#	R/scale-.R
#	tests/testthat/test-scales.R

teunbrand committed:
adopt latest palette mechanism

teunbrand committed:
make `scale_params` an S7 class

teunbrand committed:
use `update_ggplot()` S7 method instead of `ggplot_add.scale_params`
